### PR TITLE
Refactor seq2seq logic and compute PrivacyReportMetadata

### PR DESCRIPTION
**Pull request by amad-person**
Changes made:

- Refactoring seq2seq logic and tests into separate files.
- Populating PrivacyReportMetadata for seq2seq models during attacker data creation.

### CODE REVIEW

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py [hunk 1]
```diff
@@ -160,28 +193,67 @@ def _get_ranks_for_sequence(logits: np.ndarray,
...
+
+  Args:
+    batch_logits: Logits returned by a seq2seq model, dim = (num_sequences,
+    num_tokens, vocab_size).
+    batch_labels: Target labels for the seq2seq model, dim = (num_sequences,
+    num_tokens, 1).
+    loss: SparseCategoricalCrossentropy loss metric.
+  """
+  for sequence_logits, sequence_labels in zip(batch_logits, batch_labels):
+    loss.update_state(sequence_labels.astype(np.float32),
```
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:209]: This is currently modifying one of the input parameters. While this looks good, it can easily lead to other unintended side effects. Can we refactor this to return the new updated loss (like in Pandas or numpy, they always return the result and you need to reassign it). Sam for update_batch_accuracy

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py [hunk 2]
```diff
@@ -204,18 +276,24 @@ def create_seq2seq_attacker_data(attack_input_data: Seq2SeqAttackInputData,
     model_selection.train_test_split(
         features_all, labels_all, test_size=test_fraction, stratify=labels_all)
 
+  # Populate fields of privacy report metadata
+  privacy_report_metadata.loss_train = loss_train
```
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:280]: Can this be None?

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py (hunk 1 above)
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:209]: Sure, I will update both the methods

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py (hunk 2 above)
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:280]: Did you mean if privacy_report_metadata can be None? In line [313](https://github.com/tensorflow/privacy/pull/144/files/46bee91cda7e68f88a6336d9f299641eaba818f9..eb215072bcc18f202c69ed51508185e9bfa9744e#diff-acd940c715bb3abd65068035e671ca82e8a6e1287082ce7e1a638728f417d383R313) I perform a check for this and initialize privacy_report_metadata if it is None. I can move it to this line if it would make the code more readable

Comment (CdavM) [reply tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:280]: I see. what confuses me is that the parameter in this method has a default value of None, but there are no further checks after that. So maybe initialize to a default of PrivacyReportMetadata? or throw an error if it's None? Or move the check.

Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:280]: Oh, I can see how that is confusing. I will add a check

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia_test.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+                                 vocab_size):
+  num_tokens_in_sequence = np.random.choice(max_tokens_in_sequence,
+                                            num_sequences) + 1
+  batch_logits, batch_labels = [], []
+  for num_tokens in num_tokens_in_sequence:
+    logits, labels = _get_sequence_logits_and_labels(num_tokens, vocab_size)
+    batch_logits.append(logits)
+    batch_labels.append(labels)
+  return np.array(
+      batch_logits, dtype=object), np.array(
```
[the remark this answers is not in the document]
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia_test.py:251]: This was done because numpy was throwing a warning about creating a 2D array with variable length arrays inside it (because a batch will have sequences of diff lengths)

File: tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+      ((np.zeros(len(attack_input_train))), (np.ones(len(attack_input_test)))))
+
+  # Perform a train-test split
+  features_train, features_test, \
+  is_training_labels_train, is_training_labels_test = \
+    model_selection.train_test_split(
+        features_all, labels_all, test_size=test_fraction, stratify=labels_all)
+
+  # Populate accuracy, loss fields in privacy report metadata
+  privacy_report_metadata = privacy_report_metadata or PrivacyReportMetadata()
```
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/seq2seq_mia.py:300]: Can we move this PrivacyReportMetadata() to the argument line?

So 

privacy_report_metadata: PrivacyReportMetadata = PrivacyReportMetadata()

instead of this line. Reassigning input arguments can lead to unintended consequences, plus it's difficult to do refactoring if we create default parameters in the function.

**Review by CdavM — APPROVED** [review]
Thanks for the PR :-) ! I think all the changes were addressed now, right? Unfortunately it's Thanksgiving season in the US so this PR will only be merged next week.

**Comment by CdavM** [comment]
Hi @amad-person ! During testing, I had to replace the "sparse_categorical_crossentropy" function with "SparseCategoricalCrossentropy" (perhaps I'm using an older version of TF that doesn't have the other function). After doing that, the test "test_run_seq2seq_attack_calculates_correct_metadata" started failing. I then noticed that the metadata results are highly dependent on the seed value. Is this expected?

**Comment by amad-person** [comment]
> Hi @amad-person ! During testing, I had to replace the "sparse_categorical_crossentropy" function with "SparseCategoricalCrossentropy" (perhaps I'm using an older version of TF that doesn't have the other function). After doing that, the test "test_run_seq2seq_attack_calculates_correct_metadata" started failing. I then noticed that the metadata results are highly dependent on the seed value. Is this expected?

I'm generating the test input using `get_seq2seq_test_input()` randomly instead of hard coding it, and this function takes in a seed value so that the generated test input is same across multiple runs (so that the metadata values returned can be checked using assert statements). 

Would an alternative approach be better to avoid the dependency on the seed value?

**Comment by CdavM** [comment]
@amad-person Thanks for clarifying. I took another look. 

The problem with random input is that it's difficult to predict what the output should be. We also rely on the fact that the underlying np.random mechanisms will never change with a newer version. We're also using random input in mia_test.py, but we only check rudimentary things like number of items (that can still be predicted).


I'd prefer using deterministic input for things like checking the correct metadata. That way, we can compute it by hand once and then be sure that it's correct. What do you think?

**Comment by amad-person** [comment]
@CdavM that makes sense. I'll change the metadata test to take deterministic input then.

**Comment by amad-person** [comment]
@CdavM I pushed the changes. Please let me know if they're alright, thanks!

**Review by CdavM — APPROVED** [review]
